Cleanup of error handling of some of the files under goto-programs

NlightNFotis opened the pull request:
This is nothing new as a PR, it's just me taking over #2836. I have now rebased and I am now responsible for further development on it.

Addressing some of the comments by @kroening on the previous PR: There should be a new PR in the next few days, where we will have structured goto_locationt inside the invariants, to my understanding at least. This PR will be extended to handle the appropriate cases once that goes in.

Commit:
Convert asserts and throws to INVARIANTs for files starting with g

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 8654b36f9778e99c4b05744bd457cd5b5d39e631).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

tautschnig commented on src/goto-programs/goto_convert_side_effect.cpp:
```diff
@@ -143,17 +138,16 @@ void goto_convertt::remove_pre(
-         statement==ID_predecrement);
+  DATA_INVARIANT(
+    statement == ID_preincrement || statement == ID_predecrement,
+    " expected preincrement or predecrement");
```
Spurious space at beginning of error message?

tautschnig commented on src/goto-programs/goto_convert_side_effect.cpp:
```diff
@@ -350,20 +341,14 @@ void goto_convertt::remove_function_call(
+  DATA_INVARIANT(
+    !expr.operands().empty(),
+    expr.find_source_location().as_string() +
+      ": function_call expects at least one operand");
```
Above we require exactly two operands - could both of these `DATA_INVARIANT`s be fused into one, at the beginning of the method?

NlightNFotis replied:
Yes, I opted to go with the stricter version (the one requiring exactly two operands) of this invariant, given that I find that both operands are being used at various points of this function.

tautschnig commented on src/goto-programs/goto_convert_side_effect.cpp:
```diff
@@ -445,7 +430,7 @@ void goto_convertt::remove_cpp_delete(
   goto_programt &dest)
 {
-  assert(expr.operands().size()==1);
+  DATA_INVARIANT(expr.operands().size() == 1, "cpp_delete expected 1 operand");
```
Let's spell out "one", for consistency. Or change all of them to numerals. There's ample inconsistency in this PR as far as this is concerned.

NlightNFotis replied:
Agreed. I will handle all of the cases in this PR.

tautschnig commented on src/goto-programs/goto_program.cpp:
```diff
@@ -163,8 +163,9 @@ std::ostream &goto_programt::output_instruction(
-                     "size of target list");
+      DATA_INVARIANT(
+        instruction.targets.size() == exception_list.size(),
+        "size of target list");
```
Hmm, what's wrong with the size?

NlightNFotis commented on src/goto-programs/goto_convert_side_effect.cpp:
```diff
@@ -235,18 +227,16 @@ void goto_convertt::remove_post(
-         statement==ID_postdecrement);
+  DATA_INVARIANT(
+    statement == ID_postincrement || statement == ID_postdecrement,
+    " expected postincrement or postdecrement");
```
Here's also a spurious space at the beginning of the error message, which I have gone ahead and removed.

NlightNFotis commented on src/goto-programs/goto_program.cpp:
```diff
@@ -235,23 +237,23 @@ void parse_lhs_read(const exprt &src, std::list<exprt> &dest)
   if(src.id()==ID_dereference)
   {
-    assert(src.operands().size()==1);
+    PRECONDITION(src.operands().size() == 1);
```
I figured this would go very well with the changes suggested in #2902, so I went ahead to change all of these, by removing the duplicate preconditions and converting the untyped expressions into the correct types with the typecast expressions.

NlightNFotis committed:
Remove some preconditions and appropriately type the expressions.

NlightNFotis commented:
@tautschnig I have fixed your comments, and I also figured out that some of the changes suggested in a previous PR of mine (#2902) were applicable here, so I removed some of the preconditions and substituted them for a finer typing of `expr`s at the appropriate places.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 612ef30046ddf7df906561b9ce668e180b1d3bdb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84857760

peterschrammel commented on src/goto-programs/goto_convert_exceptions.cpp:
```diff
@@ -18,12 +18,10 @@ void goto_convertt::convert_msc_try_finally(
-  }
+  DATA_INVARIANT(
+    code.operands().size() == 2,
+    code.find_source_location().as_string() +
```
Use invariant with diagnostics to pass the source_location (several occurrences).

peterschrammel commented on src/goto-programs/goto_convert_side_effect.cpp:
```diff
@@ -143,17 +138,16 @@ void goto_convertt::remove_pre(
-         statement==ID_predecrement);
+  DATA_INVARIANT(
+    statement == ID_preincrement || statement == ID_predecrement,
+    "expected preincrement or predecrement");
```
`expects` (several occurrences)

peterschrammel commented on src/goto-programs/goto_program.cpp:
```diff
@@ -555,10 +558,11 @@ void goto_programt::compute_target_numbers()
+          t->target_number != 0, "instruction's number cannot be zero");
+        DATA_INVARIANT(
+          t->target_number != instructiont::nil_target,
+          "instruction cannot be an invalid target");
```
Better: `GOTO instruction target cannot be nil_target`

peterschrammel commented on src/goto-programs/goto_program.cpp:
```diff
@@ -555,10 +558,11 @@ void goto_programt::compute_target_numbers()
-        DATA_INVARIANT(t->target_number!=instructiont::nil_target,
-                       "target numbers");
+        DATA_INVARIANT(
+          t->target_number != 0, "instruction's number cannot be zero");
```
Better: `GOTO instruction target cannot be zero`

tautschnig reviewed (approved):
Looks ok to me.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: c952f5561154fc69297ef62a3bc3c8861c8899a4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85323156
